File: kuna/catalog.py
"""The kuna assertion catalog: the discovery half of the LLM control API.

An external operator (a human, or an LLM agent) needs two things to drive kuna's
sub-stage assertions in production:

  1. discover what is settable, what values it takes, and when to flip it -- this
     module (``python -m kuna.catalog``), and
  2. flip them per-decompilation -- ``kuna.decompile``'s ``--option`` / ``--kassert``
     flags (which an LLM picks using this catalog).

The catalog itself lives in ONE place in the C++ source (``kuna_stages.cc``
``settableTable``) and is emitted as JSON by the ``stage catalog`` console command.
This module runs that command and parses the result, so the docs an agent reads can
never drift from the binary's actual behavior. With a binary/function it also reports
each option's LIVE value for that program.

CLI:
    python -m kuna.catalog                 # human-readable table
    python -m kuna.catalog --json          # the raw JSON array (for an LLM/tool)
    python -m kuna.catalog --markdown      # regenerate docs/assertions.md body
    python -m kuna.catalog --option NAME   # one entry
"""
import argparse
import json
import os
import subprocess
import sys

from . import paths


class CatalogError(Exception):
    """Could not obtain the assertion catalog from the decompiler."""


def catalog(*, option=None, decomp_dbg=None, sleighpath=None, timeout=60):
    """Return the kuna settable-assertion catalog as a list of dicts.

    Drives ``decomp_dbg`` with ``stage catalog`` (no binary needed -- the catalog is
    static documentation) and parses the emitted JSON. With ``option`` set, returns a
    one-element list for that option (or raises if unknown).
    """
    bin_path = paths.decomp_dbg() if decomp_dbg is None else os.path.abspath(str(decomp_dbg))
    if not os.path.exists(bin_path):
        raise CatalogError("decomp_dbg not built at %s -- run `make binaries`" % bin_path)
    specs = str(paths.specs_dir()) if sleighpath is None else os.path.abspath(str(sleighpath))

    cmd = "stage catalog" if option is None else ("stage catalog %s" % option)
    script = cmd + "\nquit\n"
    env = dict(os.environ)
    env["SLEIGHHOME"] = specs
    try:
        proc = subprocess.run(
            [bin_path, "-s", specs],
            input=script, capture_output=True, text=True, env=env, timeout=timeout,
        )
    except subprocess.TimeoutExpired:
        raise CatalogError("`stage catalog` timed out after %ss" % timeout)

    out = proc.stdout or ""
    if "Unknown settable option" in out:
        raise CatalogError("unknown option: %r" % option)
    data = _extract_json(out)
    if data is None:
        raise CatalogError(
            "could not parse `stage catalog` output; decompiler said:\n%s" % out.strip()[:2000]
        )
    if isinstance(data, dict):
        return [data]
    return data


def _extract_json(out):
    """Pull the JSON array/object out of the console transcript.

    The console wraps the output in ``[decomp]> ...`` prompts, whose ``[``/``]`` would
    defeat a naive find/rfind. Strip the prompt, then scan from the first ``[``/``{``
    and balance brackets, ignoring any inside JSON string literals, to find the end.
    """
    out = out.replace("[decomp]>", " ")	# drop the fixed console prompt (its [ ] are not JSON)
    start = -1
    for i, ch in enumerate(out):
        if ch in "[{":
            start = i
            break
    if start == -1:
        return None
    depth = 0
    in_str = False
    escape = False
    for i in range(start, len(out)):
        ch = out[i]
        if in_str:
            if escape:
                escape = False
            elif ch == "\\":
                escape = True
            elif ch == '"':
                in_str = False
            continue
        if ch == '"':
            in_str = True
        elif ch in "[{":
            depth += 1
        elif ch in "]}":
            depth -= 1
            if depth == 0:
                try:
                    return json.loads(out[start:i + 1])
                except json.JSONDecodeError:
                    return None
    return None


def check_drift(*, decomp_dbg=None, sleighpath=None):
    """Verify the catalog documents exactly the registered kuna options.

    Returns a list of human-readable problem strings (empty == OK). Catches the drift
    that breaks the LLM API silently: a kuna ArchOption added to ``options.cc`` without
    a ``settableTable`` row (undiscoverable), or a stale catalog row for a removed
    option. Reads ``options.cc`` for the ``registerOption(new OptionXxx()); // (kuna)``
    registrations and matches them to catalog entries by the option's ``name=``.
    """
    import re

    entries = catalog(decomp_dbg=decomp_dbg, sleighpath=sleighpath)
    cat = {e["option"] for e in entries}

    cpp = paths.repo_root() / "decompiler" / "cpp"
    reg = (cpp / "options.cc").read_text()
    classes = re.findall(r"registerOption\(new (Option\w+)\(\)\);\s*//\s*\(kuna\)", reg)
    # Resolve each kuna OptionXxx class to its option name= string from the kuna_*.hh.
    headers = "".join(
        p.read_text() for p in cpp.glob("kuna_*.hh")
    )
    registered = set()
    unresolved = []
    for cls in classes:
        m = re.search(re.escape(cls) + r"\(void\)\s*\{\s*name\s*=\s*\"([^\"]+)\"", headers)
        if m:
            registered.add(m.group(1))
        else:
            unresolved.append(cls)

    problems = []
    for name in sorted(registered - cat):
        problems.append("registered kuna option %r has no catalog row (undiscoverable)" % name)
    for name in sorted(cat - registered):
        problems.append("catalog row %r matches no registered kuna option (stale)" % name)
    for cls in unresolved:
        problems.append("could not resolve option name for class %s (check kuna_*.hh)" % cls)
    return problems


def to_markdown(entries):
    """Render the catalog as the body of docs/assertions.md."""
    lines = []
    lines.append("# kuna assertion catalog")
    lines.append("")
    lines.append(
        "**Generated** from the decompiler's `stage catalog` command "
        "(`python -m kuna.catalog --markdown`) -- do not edit by hand; edit "
        "`settableTable` in `decompiler/cpp/kuna_stages.cc` and regenerate."
    )
    lines.append("")
    lines.append(
        "These are the kuna stage-model sub-stage decisions an operator (human or LLM) "
        "can flip per decompilation. Defaults are the shipped values (post-DIV-2, see "
        "`docs/divergences.md`). Set any of them with "
        "`python -m kuna.decompile <bin> <fn> --option <name> <value>` (or, per "
        "function, `--kassert \"<stage> <substage> ...\"`); revert any one with its "
        "`off`/`canonical` value."
    )
    lines.append("")
    lines.append(
        "| Option | Values | Default | Stage / sub-stage | Source | Kind | Decision | When to flip |"
    )
    lines.append("|---|---|---|---|---|---|---|---|")
    for e in entries:
        vals = " \\| ".join(e["values"])
        default = e["default"] + (" ⚠️ opt-in" if e.get("destructive_as_default") else "")
        stage = "%s / %s" % (e["stage"], e["substage"])
        lines.append("| `%s` | %s | `%s` | %s | %s | %s | %s | %s |" % (
            e["option"], vals, default, stage,
            e.get("source_decompiler", ""), e.get("change_kind", ""),
            e["summary"], e["use_when"],
        ))
    lines.append("")
    lines.append("## Programmatic use")
    lines.append("")
    lines.append("```bash")
    lines.append("# discover (machine-readable):")
    lines.append("python -m kuna.catalog --json")
    lines.append("")
    lines.append("# decompile with an assertion flipped (repeatable):")
    lines.append("python -m kuna.decompile ./a.out main --option compareform canonical")
    lines.append("python -m kuna.decompile ./sparc.elf main --option returnpair single")
    lines.append("```")
    lines.append("")
    lines.append(
        "The `⚠️ opt-in` defaults (`returnpair`, `v850indirectbranch`) are "
        "documented as destructive global defaults and ship off; apply them per "
        "function / per program only (see each row's *When to flip*)."
    )
    lines.append("")
    return "\n".join(lines)


def main(argv=None):
    p = argparse.ArgumentParser(
        prog="python -m kuna.catalog",
        description="Show the kuna settable-assertion catalog (the LLM discovery surface).",
    )
    p.add_argument("--option", default=None, help="show only this option")
    g = p.add_mutually_exclusive_group()
    g.add_argument("--json", action="store_true", help="emit the raw JSON catalog")
    g.add_argument("--markdown", action="store_true", help="emit the docs/assertions.md body")
    g.add_argument("--check", action="store_true",
                   help="verify the catalog matches the registered kuna options (exit nonzero on drift)")
    p.add_argument("--decomp-dbg", default=None, help="path to the decomp_dbg binary")
    p.add_argument("--engine", choices=("cpp", "rust"), default=None,
                   help="which port's binaries to run (sets KUNA_ENGINE; default: env, else cpp)")
    p.add_argument("--sleighpath", default=None, help="SLEIGH specs root (default: <repo>/specs)")
    args = p.parse_args(argv)

    if args.engine:
        os.environ["KUNA_ENGINE"] = args.engine  # before any paths.binary() resolution

    if args.check:
        try:
            problems = check_drift(decomp_dbg=args.decomp_dbg, sleighpath=args.sleighpath)
        except CatalogError as e:
            print("error: %s" % e, file=sys.stderr)
            return 1
        if problems:
            print("catalog DRIFT:", file=sys.stderr)
            for pr in problems:
                print("  - %s" % pr, file=sys.stderr)
            return 1
        print("catalog OK: documents exactly the registered kuna options")
        return 0

    try:
        entries = catalog(option=args.option, decomp_dbg=args.decomp_dbg, sleighpath=args.sleighpath)
    except CatalogError as e:
        print("error: %s" % e, file=sys.stderr)
        return 1

    if args.json:
        print(json.dumps(entries, indent=2))
    elif args.markdown:
        print(to_markdown(entries))
    else:
        for e in entries:
            flag = "  (opt-in: destructive as global default)" if e.get("destructive_as_default") else ""
            print("%-20s %-20s default=%-10s [%s/%s]%s" % (
                e["option"], "{" + "|".join(e["values"]) + "}", e["default"],
                e["stage"], e["substage"], flag,
            ))
            print("    %s" % e["summary"])
            print("    when: %s" % e["use_when"])
            src = e.get("source_decompiler", "")
            kind = e.get("change_kind", "")
            if src or kind:
                print("    from: %s (%s)" % (src or "?", kind or "?"))
            insp = e.get("inspiration", "")
            if insp:
                print("    inspiration: %s" % insp)
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: rust/crates/kuna-cli/src/catalog.rs
//! `kuna catalog` — the Rust port of `kuna/catalog.py`.
//!
//! Drives `decomp_dbg stage catalog [<option>]`, extracts the embedded JSON from
//! the console transcript, and re-emits it.  Modes:
//!   * (default) a human-readable table,
//!   * `--json`     the catalog as `json.dumps(indent=2)` (byte-identical to the
//!                  Python `--json`),
//!   * `--markdown` the `docs/assertions.md` body (the Python `to_markdown`),
//!   * `--check`    verify the catalog documents exactly the registered kuna
//!                  options.  In the Rust-only world the old `check_drift` (which
//!                  parsed the removed `decompiler/cpp/options.cc` + `kuna_*.hh`)
//!                  can no longer run; the faithful replacement cross-checks the
//!                  catalog the binary emits against the in-process registered
//!                  option set ([`kuna_decomp::options::KUNA_OPTION_NAMES`]),
//!                  catching the same drift (a registered option with no catalog
//!                  row, or a stale catalog row).

use std::collections::BTreeSet;
use std::process::Command;

use kuna_decomp::options::KUNA_OPTION_NAMES;

use crate::jsonfmt::{dumps_indent2, extract_json_span, parse, Json};
use crate::paths;

/// Run `decomp_dbg stage catalog [<option>]` and return the parsed JSON value
/// (an array, or a single object when `option` is given).
fn run_catalog(option: Option<&str>) -> Result<Json, String> {
    let bin = paths::decomp_dbg();
    if !bin.exists() {
        return Err(format!(
            "decomp_dbg not built at {} -- run `make binaries`",
            bin.display()
        ));
    }
    let specs = paths::specs_dir();
    let cmd = match option {
        None => "stage catalog\nquit\n".to_string(),
        Some(o) => format!("stage catalog {o}\nquit\n"),
    };

    let output = Command::new(&bin)
        .arg("-s")
        .arg(&specs)
        .env("SLEIGHHOME", &specs)
        .stdin(std::process::Stdio::piped())
        .stdout(std::process::Stdio::piped())
        .stderr(std::process::Stdio::piped())
        .spawn()
        .and_then(|mut child| {
            use std::io::Write;
            if let Some(mut sin) = child.stdin.take() {
                let _ = sin.write_all(cmd.as_bytes());
            }
            child.wait_with_output()
        })
        .map_err(|e| format!("failed to run decomp_dbg: {e}"))?;

    let out = String::from_utf8_lossy(&output.stdout);
    if out.contains("Unknown settable option") {
        return Err(format!("unknown option: {:?}", option.unwrap_or("")));
    }
    let span = extract_json_span(&out).ok_or_else(|| {
        format!(
            "could not parse `stage catalog` output; decompiler said:\n{}",
            out.trim().chars().take(2000).collect::<String>()
        )
    })?;
    parse(span).ok_or_else(|| "could not parse `stage catalog` JSON".to_string())
}

/// Normalize to a `Vec` of object entries (the catalog is an array; a single
/// option comes back as one object).
fn entries(v: &Json) -> Vec<&Json> {
    match v {
        Json::Array(items) => items.iter().collect(),
        obj @ Json::Object(_) => vec![obj],
        _ => Vec::new(),
    }
}

/// Look up a string field on an object entry.
fn field<'a>(entry: &'a Json, key: &str) -> Option<&'a str> {
    if let Json::Object(pairs) = entry {
        for (k, v) in pairs {
            if k == key {
                if let Json::Str(s) = v {
                    return Some(s);
                }
            }
        }
    }
    None
}

fn bool_field(entry: &Json, key: &str) -> bool {
    if let Json::Object(pairs) = entry {
        for (k, v) in pairs {
            if k == key {
                return matches!(v, Json::Bool(true));
            }
        }
    }
    false
}

/// String array field (e.g. `values`).
fn arr_field(entry: &Json, key: &str) -> Vec<String> {
    if let Json::Object(pairs) = entry {
        for (k, v) in pairs {
            if k == key {
                if let Json::Array(items) = v {
                    return items
                        .iter()
                        .filter_map(|i| match i {
                            Json::Str(s) => Some(s.clone()),
                            _ => None,
                        })
                        .collect();
                }
            }
        }
    }
    Vec::new()
}

/// `--json`: byte-identical to `json.dumps(entries, indent=2)`.
pub fn cmd_json(option: Option<&str>) -> i32 {
    match run_catalog(option) {
        Ok(v) => {
            // Python always re-dumps the *list* form (catalog() returns a list,
            // wrapping a single dict in `[data]`).
            let list = match &v {
                Json::Array(_) => v.clone(),
                obj => Json::Array(vec![obj.clone()]),
            };
            println!("{}", dumps_indent2(&list));
            0
        }
        Err(e) => {
            eprintln!("error: {e}");
            1
        }
    }
}

/// `--markdown`: the body of `docs/assertions.md` (port of `to_markdown`).
pub fn cmd_markdown(option: Option<&str>) -> i32 {
    let v = match run_catalog(option) {
        Ok(v) => v,
        Err(e) => {
            eprintln!("error: {e}");
            return 1;
        }
    };
    let mut lines: Vec<String> = Vec::new();
    lines.push("# kuna assertion catalog".into());
    lines.push(String::new());
    lines.push(
        "**Generated** from the decompiler's `stage catalog` command \
         (`python -m kuna.catalog --markdown`) -- do not edit by hand; edit \
         `settableTable` in `decompiler/cpp/kuna_stages.cc` and regenerate."
            .into(),
    );
    lines.push(String::new());
    lines.push(
        "These are the kuna stage-model sub-stage decisions an operator (human or LLM) \
         can flip per decompilation. Defaults are the shipped values (post-DIV-2, see \
         `docs/divergences.md`). Set any of them with \
         `python -m kuna.decompile <bin> <fn> --option <name> <value>` (or, per \
         function, `--kassert \"<stage> <substage> ...\"`); revert any one with its \
         `off`/`canonical` value."
            .into(),
    );
    lines.push(String::new());
    lines.push(
        "| Option | Values | Default | Stage / sub-stage | Source | Kind | Decision | When to flip |"
            .into(),
    );
    lines.push("|---|---|---|---|---|---|---|---|".into());
    for e in entries(&v) {
        let vals = arr_field(e, "values").join(" \\| ");
        let mut default = field(e, "default").unwrap_or("").to_string();
        if bool_field(e, "destructive_as_default") {
            default.push_str(" \u{26a0}\u{fe0f} opt-in");
        }
        let stage = format!(
            "{} / {}",
            field(e, "stage").unwrap_or(""),
            field(e, "substage").unwrap_or("")
        );
        lines.push(format!(
            "| `{}` | {} | `{}` | {} | {} | {} | {} | {} |",
            field(e, "option").unwrap_or(""),
            vals,
            default,
            stage,
            field(e, "source_decompiler").unwrap_or(""),
            field(e, "change_kind").unwrap_or(""),
            field(e, "summary").unwrap_or(""),
            field(e, "use_when").unwrap_or(""),
        ));
    }
    lines.push(String::new());
    lines.push("## Programmatic use".into());
    lines.push(String::new());
    lines.push("```bash".into());
    lines.push("# discover (machine-readable):".into());
    lines.push("python -m kuna.catalog --json".into());
    lines.push(String::new());
    lines.push("# decompile with an assertion flipped (repeatable):".into());
    lines.push("python -m kuna.decompile ./a.out main --option compareform canonical".into());
    lines.push("python -m kuna.decompile ./sparc.elf main --option returnpair single".into());
    lines.push("```".into());
    lines.push(String::new());
    lines.push(
        "The `\u{26a0}\u{fe0f} opt-in` defaults (`returnpair`, `v850indirectbranch`) are \
         documented as destructive global defaults and ship off; apply them per \
         function / per program only (see each row's *When to flip*)."
            .into(),
    );
    lines.push(String::new());
    println!("{}", lines.join("\n"));
    0
}

/// (default) the human-readable table (port of catalog.main's text branch).
pub fn cmd_text(option: Option<&str>) -> i32 {
    let v = match run_catalog(option) {
        Ok(v) => v,
        Err(e) => {
            eprintln!("error: {e}");
            return 1;
        }
    };
    for e in entries(&v) {
        let flag = if bool_field(e, "destructive_as_default") {
            "  (opt-in: destructive as global default)"
        } else {
            ""
        };
        let values = format!("{{{}}}", arr_field(e, "values").join("|"));
        println!(
            "{:<20} {:<20} default={:<10} [{}/{}]{}",
            field(e, "option").unwrap_or(""),
            values,
            field(e, "default").unwrap_or(""),
            field(e, "stage").unwrap_or(""),
            field(e, "substage").unwrap_or(""),
            flag,
        );
        println!("    {}", field(e, "summary").unwrap_or(""));
        println!("    when: {}", field(e, "use_when").unwrap_or(""));
        let src = field(e, "source_decompiler").unwrap_or("");
        let kind = field(e, "change_kind").unwrap_or("");
        if !src.is_empty() || !kind.is_empty() {
            println!(
                "    from: {} ({})",
                if src.is_empty() { "?" } else { src },
                if kind.is_empty() { "?" } else { kind }
            );
        }
        let insp = field(e, "inspiration").unwrap_or("");
        if !insp.is_empty() {
            println!("    inspiration: {insp}");
        }
    }
    0
}

/// `--check`: verify the catalog documents exactly the registered kuna options.
///
/// The Rust-only replacement for `catalog.check_drift`: compare the option names
/// the binary's `stage catalog` emits against the in-process registered set
/// (`KUNA_OPTION_NAMES`).  A registered option with no catalog row is
/// undiscoverable; a catalog row matching no registered option is stale.  The
/// build-time codegen (`stages.toml` -> `SETTABLE_TABLE`) already fences the
/// catalog/table consistency at compile time; this end-to-end check fences the
/// catalog (what the binary emits) against the registration list.
pub fn cmd_check() -> i32 {
    let v = match run_catalog(None) {
        Ok(v) => v,
        Err(e) => {
            eprintln!("error: {e}");
            return 1;
        }
    };
    let cat: BTreeSet<String> = entries(&v)
        .iter()
        .filter_map(|e| field(e, "option").map(|s| s.to_string()))
        .collect();
    let registered: BTreeSet<String> = KUNA_OPTION_NAMES.iter().map(|s| s.to_string()).collect();

    let mut problems: Vec<String> = Vec::new();
    for name in registered.difference(&cat) {
        problems.push(format!(
            "registered kuna option {name:?} has no catalog row (undiscoverable)"
        ));
    }
    for name in cat.difference(&registered) {
        problems.push(format!(
            "catalog row {name:?} matches no registered kuna option (stale)"
        ));
    }

    if !problems.is_empty() {
        eprintln!("catalog DRIFT:");
        for p in &problems {
            eprintln!("  - {p}");
        }
        return 1;
    }
    println!("catalog OK: documents exactly the registered kuna options");
    0
}
